test/inductor/test_compiled_autograd.py

```diff
@@ -0,0 +1,244 @@
+# Owner(s): ["module: inductor"]
+import torch
+from torch import _inductor as inductor
+from torch._dynamo import compiled_autograd
+from torch._dynamo.test_case import run_tests, TestCase
+from torch._dynamo.utils import counters
+from torch.testing._internal.inductor_utils import HAS_CPU
+
+
+def compiler_fn(gm):
+    """Same as torch.compile() but counts number of compiles"""
+
+    def inner_compiler(gm_, example_inputs_):
+        counters["compiled_autograd"]["compiles"] += 1
+        return inductor.compile(gm_, example_inputs_)
+
+    return torch.compile(gm, backend=inner_compiler, fullgraph=True, dynamic=True)
+
+
+# TODO(jansel): hooks as lambdas creates recompiles in dynamo, we should fix that
+
+
+def hook1(grad):
+    return grad * 2
+
+
+def hook2(grads):
+    return (grads[0] + 1,)
+
+
+def hook3(gI, gO):
+    return (torch.sin(gI[0]) + gO[0],)
+
+
+class TestCompiledAutograd(TestCase):
+    def _common(self, fn, count=1):
+        with torch.autograd.set_multithreading_enabled(False):
+            torch._dynamo.reset()
+            counters["compiled_autograd"].clear()
+            torch.manual_seed(123)
+            expected = list(fn())
+            torch.manual_seed(123)
+            with compiled_autograd.enable(compiler_fn):
+                actual = list(fn())
+            self.assertEqual(expected, actual)
+            self.assertEqual(counters["compiled_autograd"]["captures"], count)
+            self.assertEqual(counters["compiled_autograd"]["compiles"], count)
+
+    def test_basic(self):
+        def fn():
+            model = torch.nn.Sequential(
+                torch.nn.Linear(4, 4),
+                torch.nn.ReLU(),
+                torch.nn.Linear(4, 4),
+                torch.nn.ReLU(),
+            )
+            x = torch.randn([1, 4])
```

Comment:
Can you change the batch size to a non-1 value? `compiled_autograd` seems to work only when batch size is 1.

Reply:
I fixed this by adding:
https://github.com/pytorch/pytorch/blob/fef8eacb6634c5a25078be22c79b40bbe2263dd8/torch/csrc/autograd/functions/accumulate_grad.cpp#L77-L79

@albanD @ezyang do you know where this sum_to_size happens in eager mode to combine gradients in the batch dimension?  I can't seem to find it.

Reply:
Whichever op doing the "expand" on the weight is the one that is doing a broadcast and the backward of that broadcast is sum_to_size.
For historical reason, the engine automatically does that part of the formula (does it as well for type promotion btw) here: https://github.com/pytorch/pytorch/blob/0433cb05961ce91474f09898220eb611570c107a/torch/csrc/autograd/engine.cpp#L820
The implementation is a bit weird to support NestedTensor for which sizes() doesn't exist but you can find it at https://github.com/pytorch/pytorch/blob/0433cb05961ce91474f09898220eb611570c107a/torch/csrc/autograd/input_metadata.h#L117